server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -272,6 +282,11 @@ public void onResponse(Void unused) {
                                 uploadMetadata(localSegmentsPostRefresh, segmentInfos, checkpoint);
                                 logger.debug("Metadata upload successful");
                                 clearStaleFilesFromLocalSegmentChecksumMap(localSegmentsPostRefresh);
+
+                                if (segmentInfos.getGeneration() > lastRemoteCommitGeneration) {
```

Comment:
lets add a assert statement here that `segmentInfos.getGeneration() == SegmentInfos.readLatestCommit(storeDirectory).getGeneration()`

Reply:
Rather, lets move the `remoteDirectory.deleteStaleSegmentsAsync(indexShard.getRemoteStoreSettings().getMinRemoteSegmentMetadataFiles());` to within `if (segmentInfos.getGeneration() > lastRemoteCommitGeneration) {`